File: results/dnd-rest-benchmark/lifecycle-runs/20260731T160058Z_claude_claude-sonnet-5_typescript-vite/src/domain/play/currency.ts
/**
 * Per-character gold balances and campaign-local character-to-character
 * transfers within a play campaign. See shared.ts for the ownership model.
 */

import { getDb } from '../../db.ts';
import type { ApiResult, JsonValue } from '../../types.ts';
import { isValidIntInRange } from '../../validation.ts';
import {
  authenticate,
  isActor,
  isApiResult,
  findCampaign,
  findMemberByCharacterId,
  requireParticipant,
  resolveCharacterOwner,
  nextTransferId,
} from './shared.ts';

export function getCharacterCurrency(
  authHeader: string | undefined,
  campaignId: string,
  characterId: string,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const forbidden = requireParticipant(db, campaign, actor, 'not a campaign member');
  if (forbidden) return forbidden;

  const member = findMemberByCharacterId(db, campaignId, characterId);
  if (isApiResult(member)) return member;

  return { status: 200, body: { character_id: characterId, gold: member.gold } };
}

export function transferCurrency(
  authHeader: string | undefined,
  campaignId: string,
  characterId: string,
  body: JsonValue,
): ApiResult {
  const actor = authenticate(authHeader);
  if (!isActor(actor)) return actor;

  const db = getDb();
  const campaign = findCampaign(db, campaignId);
  if (isApiResult(campaign)) return campaign;

  const source = findMemberByCharacterId(db, campaignId, characterId);
  if (isApiResult(source)) return source;

  const owner = resolveCharacterOwner(db, campaignId, characterId);
  if (owner !== actor.username) {
    return { status: 403, body: { error: 'only the source character owner may transfer gold' } };
  }

  const toCharacterId = body.to_character_id;
  if (typeof toCharacterId !== 'string' || toCharacterId.length === 0 || toCharacterId === characterId) {
    return { status: 400, body: { error: 'to_character_id must be a different character in the campaign' } };
  }

  const destination = findMemberByCharacterId(db, campaignId, toCharacterId);
  if (isApiResult(destination)) {
    return { status: 400, body: { error: 'to_character_id must be a different character in the campaign' } };
  }

  if (!isValidIntInRange(body.gold, 1, Number.MAX_SAFE_INTEGER)) {
    return { status: 400, body: { error: 'gold must be a positive integer' } };
  }
  const amount = body.gold as number;

  if (source.gold < amount) {
    return { status: 409, body: { error: 'insufficient gold' } };
  }

  const fromGold = source.gold - amount;
  const toGold = destination.gold + amount;

  db.prepare('UPDATE play_campaign_members SET gold = ? WHERE campaign_id = ? AND character_id = ?').run(
    fromGold,
    campaignId,
    characterId,
  );
  db.prepare('UPDATE play_campaign_members SET gold = ? WHERE campaign_id = ? AND character_id = ?').run(
    toGold,
    campaignId,
    toCharacterId,
  );

  const transferId = nextTransferId(db, campaignId);
  db.prepare(
    'INSERT INTO play_campaign_currency_transfers (campaign_id, transfer_id, from_character_id, to_character_id, gold) VALUES (?, ?, ?, ?, ?)',
  ).run(campaignId, transferId, characterId, toCharacterId, amount);

  return {
    status: 201,
    body: {
      from_character_id: characterId,
      to_character_id: toCharacterId,
      gold: amount,
      from_gold: fromGold,
      to_gold: toGold,
      transfer_id: transferId,
    },
  };
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/routes/play/currency.rb
# Per-character gold balances and campaign-local transfers for the
# /v1/play surface.

get '/v1/play/campaigns/:id/characters/:char_id/currency' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  require_play_participant!(campaign, user, 'not a campaign member')

  find_play_member_by_character!(campaign['id'], params[:char_id])

  gold = db.execute(
    'SELECT gold FROM play_character_gold WHERE campaign_id = ? AND character_id = ?',
    [campaign['id'], params[:char_id]]
  ).first

  { character_id: params[:char_id], gold: gold ? gold['gold'] : 0 }.to_json
end

post '/v1/play/campaigns/:id/characters/:char_id/currency/transfers' do
  user = authenticate_play_request!

  campaign = find_play_campaign!(params[:id])
  from_member = find_play_member_by_character!(campaign['id'], params[:char_id])
  owner = play_character_owner(campaign['id'], params[:char_id], from_member)

  halt 403, { error: 'not the character owner' }.to_json unless owner == user['username']

  body = json_body
  halt 400, { error: 'invalid json' }.to_json if body.nil?

  to_character_id = body['to_character_id']
  gold = body['gold']

  halt 400, { error: 'invalid to_character_id' }.to_json unless to_character_id.is_a?(String) && !to_character_id.empty?
  halt 400, { error: 'cannot transfer to self' }.to_json if to_character_id == params[:char_id]

  to_member = db.execute(
    'SELECT * FROM play_campaign_members WHERE campaign_id = ? AND character_id = ?',
    [campaign['id'], to_character_id]
  ).first
  halt 400, { error: 'unknown destination character' }.to_json unless to_member

  halt 400, { error: 'invalid gold' }.to_json unless integerish(gold) && gold.to_i.positive?
  gold = gold.to_i

  from_gold_row = db.execute(
    'SELECT gold FROM play_character_gold WHERE campaign_id = ? AND character_id = ?',
    [campaign['id'], params[:char_id]]
  ).first
  from_gold = from_gold_row ? from_gold_row['gold'] : 0

  halt 409, { error: 'insufficient gold' }.to_json if gold > from_gold

  to_gold_row = db.execute(
    'SELECT gold FROM play_character_gold WHERE campaign_id = ? AND character_id = ?',
    [campaign['id'], to_character_id]
  ).first
  to_gold = to_gold_row ? to_gold_row['gold'] : 0

  from_gold_after = from_gold - gold
  to_gold_after = to_gold + gold

  db.execute(
    'INSERT INTO play_character_gold (campaign_id, character_id, gold) VALUES (?, ?, ?) ' \
    'ON CONFLICT (campaign_id, character_id) DO UPDATE SET gold = excluded.gold',
    [campaign['id'], params[:char_id], from_gold_after]
  )
  db.execute(
    'INSERT INTO play_character_gold (campaign_id, character_id, gold) VALUES (?, ?, ?) ' \
    'ON CONFLICT (campaign_id, character_id) DO UPDATE SET gold = excluded.gold',
    [campaign['id'], to_character_id, to_gold_after]
  )

  transfer_id = db.execute(
    'SELECT COALESCE(MAX(transfer_id), 0) + 1 AS n FROM play_gold_transfers WHERE campaign_id = ?',
    [campaign['id']]
  ).first['n']

  db.execute(
    'INSERT INTO play_gold_transfers (campaign_id, transfer_id, from_character_id, to_character_id, gold) VALUES (?, ?, ?, ?, ?)',
    [campaign['id'], transfer_id, params[:char_id], to_character_id, gold]
  )

  status 201
  {
    from_character_id: params[:char_id],
    to_character_id: to_character_id,
    gold: gold,
    from_gold: from_gold_after,
    to_gold: to_gold_after,
    transfer_id: transfer_id
  }.to_json
end
